ecl/eclagent/eclgraph.cpp

```diff
@@ -853,19 +854,42 @@ void EclSubGraph::updateProgress()
 {
     if (!isChildGraph && agent->queryRemoteWorkunit())
     {
-        Owned<IWUGraphStats> progress = parent.updateStats(queryStatisticsComponentType(), queryStatisticsComponentName(), id);
+        Owned<IWUGraphStats> progress = parent.updateStats(queryStatisticsComponentType(), queryStatisticsComponentName(), agent->getWorkflowId(), id);
         IStatisticGatherer & stats = progress->queryStatsBuilder();
         updateProgress(stats);
+
+        if (startGraphTime || elapsedGraphCycles)
+        {
+            WorkunitUpdate lockedwu(agent->updateWorkUnit());
+            StringBuffer subgraphid;
+            subgraphid.append(parent.queryGraphName()).append(":").append(SubGraphScopePrefix).append(id);
+            if (startGraphTime)
+                parent.updateWUStatistic(lockedwu, SSTsubgraph, subgraphid, StWhenStarted, nullptr, startGraphTime);
+            if (elapsedGraphCycles)
+                parent.updateWUStatistic(lockedwu, SSTsubgraph, subgraphid, StTimeElapsed, nullptr, cycle_to_nanosec(elapsedGraphCycles));
+        }
     }
 }
 
 void EclSubGraph::updateProgress(IStatisticGatherer &progress)
 {
-    StatsSubgraphScope subgraph(progress, id);
-    if (startGraphTime)
-        progress.addStatistic(StWhenStarted, startGraphTime);
-    if (elapsedGraphCycles)
-        progress.addStatistic(StTimeElapsed, cycle_to_nanosec(elapsedGraphCycles));
+    OwnedPtr<StatsScopeBlock> subGraph;
+    OwnedPtr<StatsScopeBlock> activityScope;
+    EclGraphElement * parentActivity = idToActivity(parentActivityId);
+    if (!parentActivity ||(parentActivity && parentActivity->loopId==0))
```

Comment:
I don't really understand this logic - it could at least do with a comment.  I also think it is probably wrong.

Currently for loopall.ecl it is generating things like:
<scope scope='w2:graph1:sg10:a18:e15_0' type='edge'/>
which is not correct.
It should be
w2:graph2:sg10:ac12:c13:sg14:e15_0

It should not be based on parentActivity->loopId, but on whether loopBody is set on the graph.  If it is it should be treated the same as the (isChildGraph) branch.

Reply:
Code modifed to use loopBody.